tests/tests/util/runner.rs

```diff
@@ -22,33 +22,35 @@ use std::time::Duration;
 
 pub struct TestAudioBackend {
     mixer: AudioMixer,
+    buffer: Vec<f32>,
 }
 
 impl TestAudioBackend {
-    pub fn new() -> Self {
-        const NUM_CHANNELS: u8 = 2;
-        const SAMPLE_RATE: u32 = 44100;
-        const BUFFER_SIZE: usize = 1024;
-
-        let mixer = AudioMixer::new(NUM_CHANNELS, SAMPLE_RATE);
-        let mixer_proxy = mixer.proxy();
-
-        std::thread::spawn(move || loop {
-            let mut buffer = [0f32; NUM_CHANNELS as usize * BUFFER_SIZE];
-            mixer_proxy.mix::<f32>(&mut buffer);
-            std::thread::sleep(std::time::Duration::from_secs_f32(
-                BUFFER_SIZE as f32 / SAMPLE_RATE as f32,
-            ));
-        });
+    const NUM_CHANNELS: u8 = 2;
+    const SAMPLE_RATE: u32 = 44100;
 
-        Self { mixer }
+    pub fn new() -> Self {
+        Self {
+            mixer: AudioMixer::new(Self::NUM_CHANNELS, Self::SAMPLE_RATE),
+            buffer: vec![],
+        }
     }
 }
 
 impl AudioBackend for TestAudioBackend {
     impl_audio_mixer_backend!(mixer);
     fn play(&mut self) {}
     fn pause(&mut self) {}
+
+    fn set_frame_rate(&mut self, frame_rate: f64) {
+        let new_buffer_size =
+            ((Self::NUM_CHANNELS as u32 * Self::SAMPLE_RATE) as f64 / frame_rate).round() as usize;
```

Comment:
Nit: I do not understand this part of the codebase, but `Self::SAMPLE_RATE` seems a bit big. I do not know what kind of values that `frame_rate` can have, and what the buffer size ought to be, but, should the buffer size increase proportionally with `Self::SAMPLE_RATE`?

Reply:
44100 is a fairly typical audio sample rate. It's almost a universal default. That's what CD audio used, and what many sound cards still use.
Frame_rate is typically somewhere around 24..60. The buffer is just there so we can tell the mixer to advance the audio stream, we don't do anything with it's contents in this test backend. And since a buffer in this setting holds all the audio data between consecutive frames, yes it should scale with the sample rate.

Reply:
That makes sense, thank you.